system/lib/pthread/pthread_create.c

```diff
@@ -109,14 +109,15 @@ void _emscripten_thread_exit(void* result) {
   emscripten_builtin_free(self->tsd);
   self->tsd = NULL;
 
+  // Not hosting a pthread anymore in this worker set __pthread_self to NULL
+  _emscripten_thread_init(0, 0, 0);
+
   // Mark the thread as no longer running.
-  // When we publish this, the main thread is free to deallocate the thread object and we are done.
+  // Once we publish this, any threads that are waiting to join with us can
+  // proceed and this worker can be recycled and used to another thread.
   self->threadStatus = 1;
-
   emscripten_futex_wake(&self->threadStatus, INT_MAX); // wake all threads
 
-  // Not hosting a pthread anymore in this worker, reset the info structures to null.
-  _emscripten_thread_init(0, 0, 0); // Unregister the thread block inside the wasm module.
   __pthread_exit_done();
```

Comment:
Should __pthread_exit_done() be moved above the futex_wake as well?

Reply:
I did think about that.

I think it probably dosn't make any difference because `__pthread_exit_done` is only used to send the `exit` message back to the main thread and this only has an effect for threads that are detached and not joinable anymore. 

So we basically have tow different modes which are mutually exclusive:

1. Detached: (__pthread_exit_done causes the main thread to collect and recycle worker)
2. Attached/Joinable: `self->threadStatus` means a waiting thread will be able to join with us recycle the worker).

So both/either of these can/will cause worker recycling and there are they both need be that last thing we do.. but only one of them we actually do anything.

I think we should maybe add some more comments to check the detach state here and only take one action... but I would want to make sure there were not races with setting the detach state before doing that.